arbor/morph/embed_pwlin.cpp

```diff
@@ -243,55 +270,44 @@ embed_pwlin::embed_pwlin(const arb::morphology& m) {
             segment_cables_[seg.id] = mcable{bid, pos0, pos1};
         }
 
-
         double length_0 = parent==mnpos? 0: data_->length[parent].back().second[1];
         data_->length[bid].push_back(0., 1, rat_element<1, 0>(length_0, length_0+branch_length));
 
         double area_0 = parent==mnpos? 0: data_->area[parent].back().second[2];
         double ixa_0 = parent==mnpos? 0: data_->ixa[parent].back().second[2];
 
-        if (length_scale==0) {
-            // Zero-length branch? Weird, but make best show of it.
-            auto& s = segments.front().prox;
-            double r = s.radius;
-            double z = s.z;
-            data_->radius[bid].push_back(0., 1., rat_element<1, 0>(r, r));
-            data_->directed_projection[bid].push_back(0., 1., rat_element<1, 0>(z-proj_shift, z-proj_shift));
-            data_->area[bid].push_back(0., 1., rat_element<2, 0>(area_0, area_0, area_0));
-            data_->ixa[bid].push_back(0., 1., rat_element<1, 1>(ixa_0, ixa_0, ixa_0));
+        for (auto i: util::count_along(segments)) {
+            auto prox = segments[i].prox;
+            auto dist = segments[i].dist;
+
+            double p0 = seg_pos[i];
+            double p1 = seg_pos[i+1];
+
+            double z0 = prox.z - proj_shift;
+            double z1 = dist.z - proj_shift;
+            data_->directed_projection[bid].push_back(p0, p1, rat_element<1, 0>(z0, z1));
+
+            double r0 = prox.radius;
+            double r1 = dist.radius;
+            data_->radius[bid].push_back(p0, p1, rat_element<1, 0>(r0, r1));
+
+            double dx = (p1-p0)*branch_length;
+            double dr = r1-r0;
+            double c = pi*std::sqrt(dr*dr+dx*dx);
+            double area_half = area_0 + (0.75*r0+0.25*r1)*c;
+            double area_1 = area_0 + (r0+r1)*c;
+            data_->area[bid].push_back(p0, p1, rat_element<2, 0>(area_0, area_half, area_1));
+            area_0 = area_1;
+
+            // (Test for positive dx explicitly in case r0 is zero.)
+            double ixa_half = ixa_0 + (dx>0? dx/(pi*r0*(r0+r1)): 0);
```

Comment:
The test is there to avoid a 0/0 evaluation and consequent NaN; an infinity is not great either, as we'll get a NaN when taking the difference after a radius zero point, but avoiding it here gives us a chance to catch that case later, if we want to.

Reply:
Okay.